File: a_Easy/1_twoSum/twosum.py
"""This is a function that adds two numbers from a list to hit a target
"""

def twoSum(nums: list[int], target: int):
    if len(nums) < 2:
        print("Invalid: initial if length check")
        return [None, None]
    elif nums[0] + nums[1] == target:
        print("Hit the target: initial elif first two indexes work")
        return [0, 1]
    
    ans = [nums[0], nums[1]]
    print(f"Initial assignment ans = [nums[0],nums[1]] {ans}")
    c = 0
    # d = 1
    # print("Initial assignment c = 0 d = 1")
    print("Initial assignment c = 0")
    
    print("Starting c-loop")
    while c < len(nums):
        d = c + 1
        print(f"(c-loop) c = {c} Assign d one more than c: d = {d}")

        # print(f"Checking if c:{c} == d:{d}")
        # if c == d:
        #     print("c was equal to d (c-loop execution) increment c and continue loop")
        #     c += 1
        #     continue
        # else:
        print(f"Starting d-loop")
        while d < len(nums):
            if nums[c] + nums[d] == target:
                print(f"(d-loop) checking nums[c] + nums[d] against target success: c = {c} d = {d} target = {target}")
                print(f"(d-loop) Assign ans[0, 1] to {nums[c]},{nums[d]}")
                ans[0], ans[1] = nums[c], nums[d]
                print(ans)
                return [c, d]
            else:
                print(f"(d-loop) nums[{c}]:{nums[c]} + nums[{d}]:{nums[d]} didn't equal {target}")
                d += 1
                print(f"(d-loop) Increment d to {d}")
        
        print(f"(c-loop) d-loop @ c = {c} completed for length of list and d = {d}")
        print(f"(c-loop) c = {c} Increment c and loop again")
        c += 1

File: a_Easy/1_twoSum/twosum.js
const twoSum = (numList, targetSum) => {
  let in1 = 0;
  let in2 = 1;
  console.log(`index 1: ${in1}, index 2: ${in2}`);

  if (numList.length < 2) {
    console.log("The array is not long enough.");
    return [false];
  } else if (numList[in1] + numList[in2] === targetSum) {
    console.log("The first two array values work.");
    console.log(`index 1: ${numList[in1]}, index 2: ${numList[in2]}`);
    return [in1, in2];
  } else {
    in2++;
    console.log(`index 1: ${in1}, index 2: ${in2}`);
  }

  console.log("first two conditionals failed");

  let testSum = 0;
  let countA = 1;
  let countB = 1;

  // Loop A
  while (in1 < numList.length) {
    console.log(`Loop A: pass ${countA}`);
    countA = countA + 1;
    console.log(`index 1: ${numList[in1]}, index 2: ${numList[in2]}`);
    console.log(`the testSum is ${testSum}`);
    console.log(`Value 1: ${numList[in1]}, Value 2: ${numList[in2]}`);
    console.log(`index 1: ${numList[in1]}, index 2: ${numList[in2]}`);

    if (testSum !== targetSum) {
      console.log("testSum wasn't the target");
      // Loop B
      while (in2 < numList.length) {
        console.log(`Loop B: pass ${countB}`);
        countB++;

        in2++;
        testSum = numList[in1] + numList[in2];
        if (testSum === targetSum) {
          break;
        }
      }
    }
  }

  console.log(`Ans1 = ${in1} and Ans2 = ${in2}`);
  return [in1, in2];
};

let numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const target = 8;

console.log("Numbers = " + numbers);
console.log("Target = " + target);
console.log(twoSum(numbers, target));
